_posts/2018-11-01-http-part6.markdown

```diff
@@ -0,0 +1,514 @@
+        val maxConnectionsToDestination: Int = 10,
+        val connectionTimeout: Duration = Duration.ofMillis(500),
+        val socketTimeout: Duration = Duration.ofMillis(200),
+        val connectionRequestTimeout: Duration = Duration.ofSeconds(1)
+) : Client {
+
+    private val connectionPool = ClientConnectionPool(keepAliveTimeout, checkKeepAliveInterval,
+            maxConnectionsToDestination, connectionTimeout, socketTimeout, connectionRequestTimeout)
+
+    companion object {
+        const val DEFAULT_HTTP_PORT = 80
+        const val SUPPORTED_PROTOCOL = "http"
+
+        private val logger = LoggerFactory.getLogger(SocketClient::class.java)
+    }
+
+    init {
+        connectionPool.start()
+    }
+
+    override fun exchange(url: String, request: Request): Response {
+        val parsedUrl = URL(url)
+        if (parsedUrl.protocol != SUPPORTED_PROTOCOL) {
+            throw SocketClientException("${parsedUrl.protocol} is not supported. " +
+                    "Only $SUPPORTED_PROTOCOL is supported")
+        }
+
+        val connection = connectionPool.retrieveConnection(ConnectionDestination( // (2)
+                host = parsedUrl.host,
+                port = if (parsedUrl.port == -1) DEFAULT_HTTP_PORT else parsedUrl.port))
+        try {
+            return exchange(connection, request)
+        } catch (e: Exception) {
+            connection.close() // (4)
+            throw SocketClientException("Error while exchanging data", e)
+        } finally {
+            connectionPool.releaseConnection(connection) // (5)
+        }
+    }
+
+    private fun exchange(connection: Connection, request: Request): Response { // (3)
+        val input = connection.getInputStream().bufferedReader()
+        val output = PrintWriter(connection.getOutputStream())
+
+        val requestRaw = request.toRaw()
+        output.print(requestRaw)
+        output.flush()
+
+        return Response.fromRaw(input)
+    }
+
+    override fun close() {
+        connectionPool.close()
+    }
+
+    fun stats() = SocketClientStats(connectionPool.stats())
+}
+
+open class SocketClientException(msg: String, throwable: Throwable? = null): RuntimeException(msg, throwable)
+data class SocketClientStats(val poolStats: ConnectionClientPoolStats)
+```
+
+We pass a bunch of arguments that were explained above. The `checkKeepAliveInterval`  (1) is how often the checker for cleaning up the old connections runs.  We pass it to the connection pool that we start on class initialization.
+
+In the `exchange()` function, we retrieve (lease or create) a connection from the pool (2). Then we do I/O converting data between strings and our structures (3). If we catch an exception, we have to close the connection (4). It probably means that there was a violation of the HTTP protocol, therefore we cannot continue sending more data on this connection. Whether we succeed or not, we should release the connection back to the pool (5).
```

Comment:
Then we do I/O operations. I/O is input / output. `Then we do input output` doesn't sound right. We can do `input output operations`.

Po polsku też nie robimy wejścia wyjścia, tylko operacje wejścia wyjścia ( http://edu.pjwstk.edu.pl/wyklady/pro/scb/PRG2CPP_files/node101.html )